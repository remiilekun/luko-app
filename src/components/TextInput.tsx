import { useState } from "react";
import {
  TextInputProps as RNTextInputProps,
  TextInput as RNTextInput,
  Text,
  StyleSheet,
  TextStyle,
  View,
  ViewStyle,
} from "react-native";
import { colors } from "../theme/colors";
import { fonts } from "../theme/fonts";

export interface TextInputProps extends RNTextInputProps {
  addon?: JSX.Element;
  label?: string;
  margined?: boolean;
  textArea?: boolean;
  error?: string;
}

export default function TextInput({
  addon,
  error,
  label,
  margined,
  textArea,
  ...props
}: TextInputProps) {
  const [isFocused, setIsFocused] = useState(false);

  return (
    <View style={[styles.wrapper, margined ? styles.wrapperMargined : {}]}>
      {label ? <Text style={styles.label}>{label}</Text> : null}
      <View style={styles.inputWrapper}>
        <RNTextInput
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          placeholderTextColor={colors.mainGrey}
          style={[
            styles.input,
            textArea ? styles.textArea : {},
            addon ? styles.inputWithAddon : {},
            isFocused ? styles.inputFocused : {},
            Boolean(error) ? styles.inputError : {},
          ]}
          {...props}
        />

        {addon ? <View style={styles.inputAddon}>{addon}</View> : null}
      </View>
      {error ? <Text style={styles.errorText}>{error}</Text> : null}
    </View>
  );
}

const styles = StyleSheet.create({
  wrapper: {},
  wrapperMargined: {
    marginBottom: 20,
  },
  label: {
    color: colors.darkGrey,
    fontFamily: fonts.regular,
    fontSize: 13,
    fontWeight: "500",
    lineHeight: 17,
    marginBottom: 5,
  } as TextStyle,
  inputWrapper: {
    position: "relative",
  },
  input: {
    backgroundColor: colors.white,
    borderColor: colors.lightGrey,
    borderRadius: 10,
    borderStyle: "solid",
    borderWidth: 2,
    fontFamily: fonts.regular,
    fontSize: 17,
    fontWeight: "400",
    lineHeight: 20,
    paddingHorizontal: 15,
    paddingVertical: 12,
  } as ViewStyle,
  inputFocused: {
    borderColor: colors.mainBlue,
    shadowColor: "#D6E3FD",
    shadowOffset: { height: 0, width: 0 },
    shadowOpacity: 0.3,
    shadowRadius: 4,
  } as ViewStyle,
  inputError: {
    borderColor: colors.red,
    shadowColor: colors.red,
  },
  errorText: {
    color: colors.red,
    fontFamily: fonts.regular,
    fontSize: 10,
    fontWeight: "400",
    marginTop: 5,
  } as TextStyle,
  inputWithAddon: {
    paddingRight: 30,
  } as ViewStyle,
  inputAddon: {
    position: "absolute",
    right: 15,
    top: 12,
  } as ViewStyle,
  textArea: {
    height: 128,
  } as ViewStyle,
});
