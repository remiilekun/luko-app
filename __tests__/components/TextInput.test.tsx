import "react-native";
import { Text } from "react-native";
import React from "react";
import { render } from "@testing-library/react-native";
import TextInput from "../../src/components/TextInput";

describe("components", () => {
  describe("TextInput", () => {
    it("renders correctly", async () => {
      const { toJSON } = render(<TextInput />);
      expect(toJSON()).toMatchSnapshot();
    });

    it("renders label correctly", async () => {
      const { getByText } = render(<TextInput label="Label" />);
      expect(getByText("Label")).not.toBeNull();
    });

    it("renders addon correctly", async () => {
      const { getByText } = render(<TextInput addon={<Text>addon</Text>} />);
      expect(getByText("addon")).not.toBeNull();
    });

    it("renders error correctly", async () => {
      const errorMessage = "Random error message";
      const { getByText } = render(<TextInput error={errorMessage} />);
      expect(getByText(errorMessage)).not.toBeNull();
    });

    it("has correct value", async () => {
      const testValue = "test value";
      const { getByTestId } = render(<TextInput value={testValue} />);
      const input = getByTestId("input");
      expect(input.props.value).toBe(testValue);
    });
  });
});
