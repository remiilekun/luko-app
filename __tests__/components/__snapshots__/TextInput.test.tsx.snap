// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components TextInput renders correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {},
    ]
  }
  testID="wrapper"
>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#C0BEB8"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#EAE9E3",
            "borderRadius": 10,
            "borderStyle": "solid",
            "borderWidth": 2,
            "fontFamily": "Circular-Regular",
            "fontSize": 17,
            "fontWeight": "400",
            "lineHeight": 20,
            "paddingHorizontal": 15,
            "paddingVertical": 12,
          },
          Object {},
          Object {},
          Object {},
          Object {},
        ]
      }
      testID="input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
